stdlib/src/utils/static_tuple.mojo

```diff
@@ -287,6 +287,39 @@ struct InlineArray[
         ]()
         self._array = __mlir_op.`kgen.undef`[_type = Self.type]()
 
+    fn __init__(
+        inout self,
+        *,
+        owned unsafe_assume_initialized: InlineArray[
+            UnsafeMaybeUninitialized[Self.ElementType], Self.size
+        ],
+    ):
+        """Constructs an `InlineArray` from an `InlineArray` of `UnsafeMaybeUninitialized`.
+
+        Calling this function assumes that all elements in the input array are initialized.
+
+        It means that when copying, moving and destroying the new `InlineArray`, the copy constructor,
+        move constructor and destructor of the elements will be called.
+
+        If the elements of the input array are not initialized, the behavior is undefined.
+
+        There is only one situation when it is somewhat correct to use this constructor and the input array
+        do not have initialized elements. This is when the input array elements' type
+        has trivial copy constructor, move constructor and destructor (like integers or floats).
+        In this case, the values in the array will be random (since it's uninitialized memory)
+        and it's up to the user to handle them accordingly.
```

Comment:
I think this paragraph is not quite correct. The issue is subtle, but the basic problem is that _even if_ `ElementType` is valid for _every possible bit pattern_, its still undefined behavior to attempt a read of uninitialized memory, even if every possible value of that uninitialized memory would be a valid instance of the type.

This "poison" value for uninitialized data is why the [`std::mem::uninitialized()`](https://doc.rust-lang.org/stable/std/mem/fn.uninitialized.html#:~:text=The%20reason%20for%20deprecation%20is%20that%20the%20function%20basically%20cannot%20be%20used%20correctly) function was deprecated in favor of `MaybeUninit` in Rust: there is no way to use it correctly, even if the programmer only uses it for "trivial" types. The undefined-behavior-causing optimizations enabled for the compiler based on the assumption that uninitialized memory is never read are indifferent to whether that memory is a trivial type or not.

So I think its important we document clearly that callers of this function MUST take care that every element HAS been written to before doing this "assume its initialized" transmutation.

Reply:
Okay I think I understand better now, it's true that I didn't take into account all kinds of optimizations the compiler could perform. Thanks for teaching me all this! I'll fix the PR.

Reply:
@ConnorGray  it's been fixed, let me know if this docstring works for you :)

Reply:
No problem at all, thank you so much for pushing forward with such energy on this work!🙂  These are some solid foundations 🧱 you're building for folks wanting to write low-level abstractions in Mojo 🔥